README.rst
=============================
pyontruck
=============================

.. image:: https://user-images.githubusercontent.com/4689706/77090966-3504a180-6a08-11ea-8cfb-7ac145d65d43.png
    :target: https://ontruck.com

|build-status| |coverage| |docs| |license|



Django extended by Ontruck-ers

Documentation
-------------

The full documentation is at https://pyontruck.readthedocs.io.

Quickstart
----------

Install pyontruck::

    pip install pyontruck


Features
------------

* Context manager to edit dict
* Retry decorator
* Sequence utils
* Time utils

Running Tests
-------------

Does the code actually work?

Prepare test env

::

    source <YOURVIRTUALENV>/bin/activate
    (myenv) $ pip install -r requirements/test.txt


Run tests

::

    (myenv) $ make test

Run tests in several python versions using tox

::

    (myenv) $ make test-all


Run tests getting code coverage


::

    (myenv) $ make coverage


Generate documentation
----------------------

::

    (myenv) $ pip install -r requirements/dev.txt
    (myenv) $ make docs


.. |build-status| image:: https://travis-ci.org/ontruck/pyontruck.svg?branch=master
    :target: https://travis-ci.org/ontruck/pyontruck
    :alt: Build status

.. |coverage| image:: https://codecov.io/gh/ontruck/pyontruck/branch/master/graph/badge.svg
    :target: https://codecov.io/gh/ontruck/pyontruck
    :alt: Coverage status

.. |docs| image:: https://readthedocs.org/projects/pyontruck/badge/?version=latest
    :target: https://pyontruck.readthedocs.io/en/latest/?badge=latest
    :alt: Documentation Status

.. |license| image:: https://img.shields.io/pypi/l/celery.svg
    :alt: BSD License
    :target: https://opensource.org/licenses/BSD-3-Clause


HISTORY.rst
.. :changelog:

History
-------

0.1.0 (2020-03-27)
---------------------

* First release. Ready to be integrated with ontruck core


tests/test_timing.py
"""
test_pyontruck
----------------------------------

Tests for `pyontruck.timing` module.
"""
from datetime import datetime, timedelta
from freezegun import freeze_time
import pytest
import pytz
from pytest_cases import pytest_parametrize_plus, fixture_ref

from pyontruck.timing import get_next_monday, get_naive_time_utc, parse_to_timedelta



@pytest.fixture
@freeze_time('2020-01-30')
def get_frozen_datetime():
    return datetime.now()


@freeze_time('2020-01-30')
def test_get_next_monday_from_faked_present_moment():
    next_monday = get_next_monday()
    assert "2020-02-03" == next_monday.strftime("%Y-%m-%d")


def test_get_next_monday_from_specific_date():
    base_date = datetime(1980, 4, 29)
    next_monday = get_next_monday(base_date)
    assert "1980-05-05" == next_monday.strftime("%Y-%m-%d")


@pytest_parametrize_plus("datetime_value, country_code, exception_raised", [
    (datetime.now(), 'XX', KeyError),
    (1, 'ES', AttributeError)
])
def test_naive_time_utc_errors(datetime_value, country_code, exception_raised):
    with pytest.raises(exception_raised):
        get_naive_time_utc(datetime_value, country_code=country_code)


@pytest_parametrize_plus("datetime_value, country_code, timezone_region", [
    (fixture_ref(get_frozen_datetime), 'ES', 'Europe/Madrid'),
    (fixture_ref(get_frozen_datetime), 'UK', 'Europe/London'),
    (fixture_ref(get_frozen_datetime), 'FR', 'Europe/Paris'),
    (fixture_ref(get_frozen_datetime), 'NL', 'Europe/Amsterdam')
])
def test_naive_time_utc(datetime_value, country_code, timezone_region):
    assert pytz.timezone(timezone_region).localize(datetime_value).astimezone(pytz.utc) == get_naive_time_utc(
        datetime_value, country_code=country_code)


@pytest.mark.parametrize("timedelta_string, expected_result", [
    ('7d', timedelta(days=7)),
    ('24h', timedelta(hours=24)),
    ('60m', timedelta(minutes=60)),
    ('120s', timedelta(seconds=120))])
def test_parse_to_timedelta(timedelta_string, expected_result):
    assert parse_to_timedelta(timedelta_string) == expected_result


def test_parse_to_timedelta_error():
    with pytest.raises(Exception):
        parse_to_timedelta('wrong string')


tests/test_sequence.py
"""
test_pyontruck
----------------------------------

Tests for `pyontruck.sequence` module.
"""
import pytest
from pytest_cases import pytest_parametrize_plus, fixture_ref
from pyontruck.sequence import remove_duplicates, clean_nones


@pytest.fixture
def list_without_duplicates():
    return [1, 2, 3, 4, 5, 'a', 'e', 'i', 'o', 'u']


@pytest.fixture
def list_with_duplicates():
    return [1, 2, 3, 4, 5, 'a', 'e', 'i', 'o', 'u', 1, 2, 3, 4, 5]


@pytest.fixture
def none_list():
    return [None, None, None]


@pytest.fixture
def list_without_duplicates_and_nones(list_without_duplicates, none_list):
    return list_without_duplicates + none_list


# If you want to use fixtures as parametrize arguments, need pytest_cases extension
# Ref: https://stackoverflow.com/questions/42228895/how-to-parametrize-a-pytest-fixture/56871701#56871701
@pytest_parametrize_plus("expected_list, list_to_test", [
    (fixture_ref(list_without_duplicates), fixture_ref(list_without_duplicates)),
    ([], fixture_ref(none_list)),
    (fixture_ref(list_without_duplicates), fixture_ref(list_without_duplicates_and_nones))
])
def test_clean_nones(expected_list, list_to_test):
    assert expected_list == clean_nones(list_to_test)


def test_clean_nones_error():
    with pytest.raises(TypeError):
        clean_nones(1)


@pytest_parametrize_plus("expected_list, list_to_test", [
    (fixture_ref(list_without_duplicates), fixture_ref(list_without_duplicates)),
    (fixture_ref(list_without_duplicates), fixture_ref(list_with_duplicates))
])
def test_remove_duplicates(expected_list, list_to_test):
    assert expected_list == remove_duplicates(list_to_test)


def test_remove_duplicates_errors():
    with pytest.raises(TypeError):
        remove_duplicates(1)


tests/test_context_manager.py
"""
test_pyontruck
----------------------------------

Tests for `pyontruck.context_manager` module.
"""
import pytest
from pyontruck.context_manager import edit_dict


@pytest.mark.parametrize("dictionary, key, value", [
    ({"foo": "bar", "xxx": "yyy"}, "xxx", "zzz"),
    ({7: 3, "xxx": [1, 2, 3]}, "xxx", "zzz"),
])
def test_edit_dict(dictionary, key, value):
    # Store the value we want to modify
    original_value = dictionary.get(key)

    # Modify dictionary inside context manager
    with edit_dict(key, value, dictionary):
        assert dictionary.get(key) == value

    assert dictionary.get(key) == original_value


travis_pypi_setup.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
"""Update encrypted deploy password in Travis config file
"""


from __future__ import print_function
import base64
import json
import os
from getpass import getpass
import yaml
from cryptography.hazmat.primitives.serialization import load_pem_public_key
from cryptography.hazmat.backends import default_backend
from cryptography.hazmat.primitives.asymmetric.padding import PKCS1v15


try:
    from urllib import urlopen
except:
    from urllib.request import urlopen


GITHUB_REPO = 'ontruck/pyontruck'
TRAVIS_CONFIG_FILE = os.path.join(
    os.path.dirname(os.path.abspath(__file__)), '.travis.yml')


def load_key(pubkey):
    """Load public RSA key, with work-around for keys using
    incorrect header/footer format.

    Read more about RSA encryption with cryptography:
    https://cryptography.io/latest/hazmat/primitives/asymmetric/rsa/
    """
    try:
        return load_pem_public_key(pubkey.encode(), default_backend())
    except ValueError:
        # workaround for https://github.com/travis-ci/travis-api/issues/196
        pubkey = pubkey.replace('BEGIN RSA', 'BEGIN').replace('END RSA', 'END')
        return load_pem_public_key(pubkey.encode(), default_backend())


def encrypt(pubkey, password):
    """Encrypt password using given RSA public key and encode it with base64.

    The encrypted password can only be decrypted by someone with the
    private key (in this case, only Travis).
    """
    key = load_key(pubkey)
    encrypted_password = key.encrypt(password, PKCS1v15())
    return base64.b64encode(encrypted_password)


def fetch_public_key(repo):
    """Download RSA public key Travis will use for this repo.

    Travis API docs: http://docs.travis-ci.com/api/#repository-keys
    """
    keyurl = 'https://api.travis-ci.org/repos/{0}/key'.format(repo)
    data = json.loads(urlopen(keyurl).read())
    if 'key' not in data:
        errmsg = "Could not find public key for repo: {}.\n".format(repo)
        errmsg += "Have you already added your GitHub repo to Travis?"
        raise ValueError(errmsg)
    return data['key']


def prepend_line(filepath, line):
    """Rewrite a file adding a line to its beginning.
    """
    with open(filepath) as f:
        lines = f.readlines()

    lines.insert(0, line)

    with open(filepath, 'w') as f:
        f.writelines(lines)


def load_yaml_config(filepath):
    with open(filepath) as f:
        return yaml.load(f)


def save_yaml_config(filepath, config):
    with open(filepath, 'w') as f:
        yaml.dump(config, f, default_flow_style=False)


def update_travis_deploy_password(encrypted_password):
    """Update the deploy section of the .travis.yml file
    to use the given encrypted password.
    """
    config = load_yaml_config(TRAVIS_CONFIG_FILE)

    config['deploy']['password'] = dict(secure=encrypted_password)

    save_yaml_config(TRAVIS_CONFIG_FILE, config)

    line = ('# This file was autogenerated and will overwrite'
            ' each time you run travis_pypi_setup.py\n')
    prepend_line(TRAVIS_CONFIG_FILE, line)


def main(args):
    public_key = fetch_public_key(args.repo)
    password = args.password or getpass('PyPI password: ')
    update_travis_deploy_password(encrypt(public_key, password))
    print("Wrote encrypted password to .travis.yml -- you're ready to deploy")


if '__main__' == __name__:
    import argparse
    parser = argparse.ArgumentParser(description=__doc__)
    parser.add_argument('--repo', default=GITHUB_REPO,
                        help='GitHub repo (default: %s)' % GITHUB_REPO)
    parser.add_argument('--password',
                        help='PyPI password (will prompt if not provided)')

    args = parser.parse_args()
    main(args)


pyontruck/context_manager.py
# _mutable is an attribute defined in some classes in Django!
# Check: https://docs.djangoproject.com/en/2.2/_modules/django/http/request/#QueryDict
# TODO: add this context manager in djontruck library, not here
"""
class MakeMutable:
    def __init__(self, unmutable):
        self.unmutable = unmutable

    def __enter__(self):
        self.unmutable._mutable = True

    def __exit__(self, *args, **kwargs):
        self.unmutable._mutable = False


def make_mutable(unmutable):
    return MakeMutable(unmutable)
"""


class EditDict:
    def __init__(self, field, value, dictionary):
        self.field = field
        self.value = value
        self.dictionary = dictionary
        self.original_value = dictionary[self.field]

    def __enter__(self):
        self.dictionary[self.field] = self.value

    def __exit__(self, *args, **kwargs):
        self.dictionary[self.field] = self.original_value


def edit_dict(field, value, dictionary):
    return EditDict(field, value, dictionary)


pyontruck/timing.py
import pytz
import re
from datetime import datetime, timedelta

ES_REGION = 'Europe/Madrid'
ES_TIMEZONE = pytz.timezone(ES_REGION)
UK_REGION = 'Europe/London'
UK_TIMEZONE = pytz.timezone(UK_REGION)
FR_REGION = 'Europe/Paris'
FR_TIMEZONE = pytz.timezone(FR_REGION)
NL_REGION = 'Europe/Amsterdam'
NL_TIMEZONE = pytz.timezone(NL_REGION)

TIMEZONES = {
    'ES': ES_TIMEZONE,
    'UK': UK_TIMEZONE,
    'FR': FR_TIMEZONE,
    'NL': NL_TIMEZONE,
}


def get_next_monday(date=None):
    now = date or datetime.utcnow()
    num_days_to_add = 7 - now.weekday()
    return now + timedelta(days=num_days_to_add)


def get_naive_time_utc(datetime_value, country_code='ES'):
    """ Localize naive data and get its utc value """
    timezone = TIMEZONES[country_code]
    return timezone.localize(datetime_value).astimezone(pytz.utc)


def parse_to_timedelta(time_val):
    """
    Given a *time_val* (string) such as '5d', returns a timedelta object
    representing the given value (e.g. timedelta(days=5)).  Accepts the
    following '<num><char>' formats:

    =========   ======= ===================
    Character   Meaning Example
    =========   ======= ===================
    s           Seconds '60s' -> 60 Seconds
    m           Minutes '5m'  -> 5 Minutes
    h           Hours   '24h' -> 24 Hours
    d           Days    '7d'  -> 7 Days
    =========   ======= ===================

    Examples::

        >>> parse_to_timedelta('7d')
        datetime.timedelta(7)
        >>> parse_to_timedelta('24h')
        datetime.timedelta(1)
        >>> parse_to_timedelta('60m')
        datetime.timedelta(0, 3600)
        >>> parse_to_timedelta('120s')
        datetime.timedelta(0, 120)
    """

    if not re.match(r"^\d+[smhd]$", time_val):
        raise Exception('time_val has incorrect format')

    num = int(time_val[:-1])
    if time_val.endswith('s'):
        return timedelta(seconds=num)
    elif time_val.endswith('m'):
        return timedelta(minutes=num)
    elif time_val.endswith('h'):
        return timedelta(hours=num)
    elif time_val.endswith('d'):
        return timedelta(days=num)
